In [ ]:
#QUESTION 1:


ANS: None of the above

In [ ]:
#QUESTION 2:


ANS: array[ : 2, 1 : ] 

In [ ]:
#QUESTION 3:


ANS: Math

In [ ]:
#QUESTION 4:


ANS: df.shape

In [ ]:
#QUESTION 5:


ANS: 'e'

In [ ]:
#QUESTION 6:


ANS: Type Error

In [ ]:
#QUESTION 7:


ANS:  x = y[1][1]

In [ ]:
#QUESTION 8:


ANS: Set and Dictionary

In [ ]:
#QUESTION 9:


ANS: 3 features, 100 observations

In [ ]:
#QUESTION 10:


ANS: c 

In [1]:
# importing the necessary python libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# importing the dataset for the quiz

quiz_data = pd.read_csv("stage_A_data.csv", encoding="ISO-8859-1")

In [3]:
# viewing the first five rows of the imported dataset

quiz_data.head()

Area Code     Area  Item Code         Item  Element Code  \
0          4  Algeria       2501   Population           511   
1          4  Algeria       2501   Population          5301   
2          4  Algeria       2901  Grand Total           664   
3          4  Algeria       2901  Grand Total           674   
4          4  Algeria       2901  Grand Total           684   

                                  Element             Unit     Y2014  \
0           Total Population - Both sexes     1000 persons  38924.00   
1                Domestic supply quantity      1000 tonnes      0.00   
2           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
3  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
4      Fat supply quantity (g/capita/day)     g/capita/day     80.06   

      Y2015     Y2016     Y2017     Y2018  
0  39728.00  40551.00  41389.00  42228.00  
1      0.00      0.00      0.00      0.00  
2   3379.00   3372.00   3341.00   3322.00  
3     94.35     94.72     92.82     91.83  
4     79.36     77.40     80.19     77.28

In [4]:
# viewing the last five rows of the dataset

quiz_data.tail()

Area Code      Area  Item Code           Item  Element Code  \
60938        181  Zimbabwe       2899  Miscellaneous          5142   
60939        181  Zimbabwe       2899  Miscellaneous           645   
60940        181  Zimbabwe       2899  Miscellaneous           664   
60941        181  Zimbabwe       2899  Miscellaneous           674   
60942        181  Zimbabwe       2899  Miscellaneous           684   

                                      Element             Unit  Y2014  Y2015  \
60938                                    Food      1000 tonnes  42.00  46.00   
60939     Food supply quantity (kg/capita/yr)               kg   3.06   3.33   
60940           Food supply (kcal/capita/day)  kcal/capita/day   3.00   4.00   
60941  Protein supply quantity (g/capita/day)     g/capita/day   0.10   0.11   
60942      Fat supply quantity (g/capita/day)     g/capita/day   0.04   0.05   

       Y2016  Y2017  Y2018  
60938  33.00  19.00  16.00  
60939   2.35   1.33   1.08  
60940   3.00   1.00   1.00  
60941   0.08   0.04   0.04  
60942   0.03   0.02   0.01

In [5]:
# viewing the shape of the dataset

quiz_data.shape

(60943, 12)

The imported data has 60943 rows and 12 columns (features)

In [6]:
# viewing the summary of the information about the imported data

quiz_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60943 entries, 0 to 60942
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Area Code     60943 non-null  int64  
 1   Area          60943 non-null  object 
 2   Item Code     60943 non-null  int64  
 3   Item          60943 non-null  object 
 4   Element Code  60943 non-null  int64  
 5   Element       60943 non-null  object 
 6   Unit          60943 non-null  object 
 7   Y2014         59354 non-null  float64
 8   Y2015         59395 non-null  float64
 9   Y2016         59408 non-null  float64
 10  Y2017         59437 non-null  float64
 11  Y2018         59507 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 5.6+ MB


The dataset consists of different data types in terms of the columns (5 floats, 3 integers and 4 objects).

From the Non-Null count, it is obvious that the dataset has some null values.

In [7]:
# checking for null values

quiz_data.isnull().sum()

Area Code          0
Area               0
Item Code          0
Item               0
Element Code       0
Element            0
Unit               0
Y2014           1589
Y2015           1548
Y2016           1535
Y2017           1506
Y2018           1436
dtype: int64

The null values are in the Y2014, Y2015, Y2016, Y2017 and Y2018 columns.

In [8]:
# checking for duplicate values in the dataset

quiz_data.duplicated().sum()

0

The dataset has no duplicate values.

## Question 11
What is the total sum of Animal Fat produced in 2014 and 2017 respectively?

Hint: Perform a groupby sum aggregation on ‘Item’

In [9]:
quiz_data.groupby('Item')[['Y2014', 'Y2017']].sum()

Y2014      Y2017
Item                                     
Alcohol, Non-Food      2403.00    2348.00
Alcoholic Beverages  102410.11   95581.06
Animal Products       11935.65   11547.65
Animal fats          209460.54  269617.53
Apples and products    9499.23   10198.90
...                        ...        ...
Vegetables, Other    155038.96  157752.59
Vegetal Products     107145.19  107655.20
Wheat and products   232670.13  240047.62
Wine                   4497.36    4178.02
Yams                 200396.96  229174.59

[119 rows x 2 columns]

In [10]:
animal_fat_2014 = quiz_data.groupby('Item')[['Y2014', 'Y2017']].sum().loc['Animal fats'][0]

animal_fat_2017 = quiz_data.groupby('Item')[['Y2014', 'Y2017']].sum().loc['Animal fats'][1]

print(f'Sum of animal fats in 2014: {round(animal_fat_2014, 2)}')

print(f'Sum of animal fats in 2017: {round(animal_fat_2017, 2)}')

Sum of animal fats in 2014: 209460.54
Sum of animal fats in 2017: 269617.53


## Question 12

What is the mean and standard deviation across the whole dataset for the year 2015 to 3 decimal places?

In [11]:
round(quiz_data['Y2015'].describe()['mean'], 3)

135.236

In [12]:
round(quiz_data['Y2015'].describe()['std'], 3)

1603.404

 Alternatively...

In [13]:
mean = quiz_data['Y2015'].mean()

std = quiz_data['Y2015'].std()

print(f'MEAN: {round(mean, 3)}')

print(f'STANDARD DEVIATION: {round(std, 3)}')

MEAN: 135.236
STANDARD DEVIATION: 1603.404


## Question 13

What is the total number and percentage of missing data in 2016 to 2 decimal places?

In [14]:
null_2016 = quiz_data['Y2016'].isnull().sum()

In [15]:
total_2016 = len(quiz_data['Y2016'])

In [16]:
null_percent = (null_2016/total_2016)*100

In [17]:
print(f'Total missing data: {null_2016} \nPercent of missing data: {round(null_percent, 2)}')

Total missing data: 1535 
Percent of missing data: 2.52


## Question 14
Which year had the highest correlation with ‘Element Code’?

In [18]:
quiz_data.corr()['Element Code']['Y2014':].sort_values(ascending=False)

Y2014    0.024457
Y2018    0.024279
Y2017    0.024254
Y2015    0.023889
Y2016    0.023444
Name: Element Code, dtype: float64

The year with the highest correlation with Element Code is **2014**.

## Question 15

What year has the highest sum of Import Quantity?

Hint- Perform a groupby operation on ‘Element’ and use the resulting Dataframe to answer the question

In [19]:
element_sum = quiz_data.groupby('Element').sum()

In [20]:
element_sum.loc['Import Quantity']['Y2014':].sort_values(ascending=False)

Y2017    294559.09
Y2018    287997.09
Y2016    286582.78
Y2014    274144.48
Y2015    267018.46
Name: Import Quantity, dtype: float64

The year with the highest sum of Import Quantity is **2017**.

## Question 16

What is the total number of the sum of Production in 2014?

Hint- Perform a groupby operation on ‘Element’ and use the resulting Dataframe to answer the question

In [21]:
quiz_data.groupby('Element').sum()

Area Code  Item Code  Element Code  \
Element                                                                      
Domestic supply quantity                   708993   14197445      28068795   
Export Quantity                            599910   11840553      26026133   
Fat supply quantity (g/capita/day)         675050   13535000       3435732   
Feed                                       176272    3538507       7282199   
Food                                       663295   13285035      25406622   
Food supply (kcal/capita/day)              674057   13511060       3329296   
Food supply quantity (kg/capita/yr)        658446   13185401       3163725   
Import Quantity                            688174   13795966      28834929   
Losses                                     274353    5424803      10292107   
Other uses (non-food)                      235554    4729749       8926728   
Processing                                 271940    5350416      10313310   
Production                                 526751   10450053      21388191   
Protein supply quantity (g/capita/day)     675050   13535000       3385502   
Residuals                                  623271   12421089      24066350   
Seed                                       103537    2035933       4211574   
Stock Variation                            571566   11329527      21464704   
Total Population - Both sexes                6020     112545         22995   
Tourist consumption                         50308    1486528       2869905   

                                             Y2014       Y2015       Y2016  \
Element                                                                      
Domestic supply quantity                1996716.35  2021493.55  2044842.70   
Export Quantity                          150020.64   157614.47   151920.46   
Fat supply quantity (g/capita/day)        10225.56    10235.74    10102.77   
Feed                                     216927.89   225050.22   228958.65   
Food                                    1212332.49  1232361.10  1247022.17   
Food supply (kcal/capita/day)            454257.00   453383.00   451810.00   
Food supply quantity (kg/capita/yr)       49650.63    49345.13    48985.28   
Import Quantity                          274144.48   267018.46   286582.78   
Losses                                   153223.00   155439.00   157787.00   
Other uses (non-food)                     78718.13    66254.41    69563.68   
Processing                               282923.00   287929.00   280631.00   
Production                              1931287.75  1947019.39  1943537.15   
Protein supply quantity (g/capita/day)    11836.46    11833.95    11779.69   
Residuals                                 30149.00    30045.00    37224.00   
Seed                                      21922.92    23976.82    23389.20   
Stock Variation                           58749.83    34910.99    33140.12   
Total Population - Both sexes           1031585.00  1058081.00  1085107.00   
Tourist consumption                         416.00      349.00       89.00   

                                             Y2017       Y2018  
Element                                                         
Domestic supply quantity                2088198.10  2161192.10  
Export Quantity                          182338.80   181594.80  
Fat supply quantity (g/capita/day)        10253.84    10258.69  
Feed                                     223705.68   233489.68  
Food                                    1258888.28  1303841.28  
Food supply (kcal/capita/day)            454681.00   455261.00  
Food supply quantity (kg/capita/yr)       48690.04    49056.85  
Import Quantity                          294559.09   287997.09  
Losses                                   160614.00   163902.00  
Other uses (non-food)                     91645.97    91300.97  
Processing                               292836.00   308429.00  
Production                              2030056.89  2075072.89  
Protein 

In [22]:
production_2014 = quiz_data.groupby('Element').sum().loc['Production']['Y2014']

print(f'Sum of Production in 2014: {round(production_2014, 2)}')

Sum of Production in 2014: 1931287.75


## Question 17

Which of these elements had the highest sum in 2018?

Hint- Select columns ‘Y2018’ and ‘Element’, Perform a groupby operation on ‘Element’ on the selected dataframe and answer the question.

In [23]:
quiz_data.groupby('Element')[['Y2018']].sum().sort_values(by='Y2018', ascending=False)

Y2018
Element                                           
Domestic supply quantity                2161192.10
Production                              2075072.89
Food                                    1303841.28
Total Population - Both sexes           1140605.00
Food supply (kcal/capita/day)            455261.00
Processing                               308429.00
Import Quantity                          287997.09
Feed                                     233489.68
Export Quantity                          181594.80
Losses                                   163902.00
Other uses (non-food)                     91300.97
Food supply quantity (kg/capita/yr)       49056.85
Residuals                                 34864.00
Seed                                      25263.14
Stock Variation                           20577.91
Protein supply quantity (g/capita/day)    11833.56
Fat supply quantity (g/capita/day)        10258.69
Tourist consumption                          90.00

The element with the highest sum in 2018 is **Domestic supply quantity**.

## Question 18

Which of these elements had the 3rd lowest sum in 2018?

Hint- Select columns ‘Y2018’ and ‘Element’, Perform a groupby operation on ‘Element’ on the selected dataframe and answer the question.

In [24]:
quiz_data.groupby('Element')[['Y2018']].sum().sort_values(by='Y2018', ascending=True)

Y2018
Element                                           
Tourist consumption                          90.00
Fat supply quantity (g/capita/day)        10258.69
Protein supply quantity (g/capita/day)    11833.56
Stock Variation                           20577.91
Seed                                      25263.14
Residuals                                 34864.00
Food supply quantity (kg/capita/yr)       49056.85
Other uses (non-food)                     91300.97
Losses                                   163902.00
Export Quantity                          181594.80
Feed                                     233489.68
Import Quantity                          287997.09
Processing                               308429.00
Food supply (kcal/capita/day)            455261.00
Total Population - Both sexes           1140605.00
Food                                    1303841.28
Production                              2075072.89
Domestic supply quantity                2161192.10

The element with the third lowest sum in 2018 is **Protein supply quantity (g/capita/day)**.

## Question 19

What is the total Import Quantity in Algeria in 2018?

In [25]:
quiz_data_copy = quiz_data.copy()

In [26]:
quiz_data_copy.set_index('Area').loc['Algeria']

Area Code  Item Code           Item  Element Code  \
Area                                                         
Algeria          4       2501     Population           511   
Algeria          4       2501     Population          5301   
Algeria          4       2901    Grand Total           664   
Algeria          4       2901    Grand Total           674   
Algeria          4       2901    Grand Total           684   
...            ...        ...            ...           ...   
Algeria          4       2899  Miscellaneous          5142   
Algeria          4       2899  Miscellaneous           645   
Algeria          4       2899  Miscellaneous           664   
Algeria          4       2899  Miscellaneous           674   
Algeria          4       2899  Miscellaneous           684   

                                        Element             Unit     Y2014  \
Area                                                                         
Algeria           Total Population - Both sexes     1000 persons  38924.00   
Algeria                Domestic supply quantity      1000 tonnes      0.00   
Algeria           Food supply (kcal/capita/day)  kcal/capita/day   3377.00   
Algeria  Protein supply quantity (g/capita/day)     g/capita/day     94.90   
Algeria      Fat supply quantity (g/capita/day)     g/capita/day     80.06   
...                                         ...              ...       ...   
Algeria                                    Food      1000 tonnes     75.00   
Algeria     Food supply quantity (kg/capita/yr)               kg      1.93   
Algeria           Food supply (kcal/capita/day)  kcal/capita/day      2.00   
Algeria  Protein supply quantity (g/capita/day)     g/capita/day      0.06   
Algeria      Fat supply quantity (g/capita/day)     g/capita/day      0.03   

            Y2015     Y2016     Y2017     Y2018  
Area                                             
Algeria  39728.00  40551.00  41389.00  42228.00  
Algeria      0.00      0.00      0.00      0.00  
Algeria   3379.00   3372.00   3341.00   3322.00  
Algeria     94.35     94.72     92.82     91.83  
Algeria     79.36     77.40     80.19     77.28  
...           ...       ...       ...       ...  
Algeria     77.00     78.00     80.00     82.00  
Algeria      1.93      1.93      1.93      1.93  
Algeria      2.00      2.00      2.00      2.00  
Algeria      0.06      0.06      0.06      0.06  
Algeria      0.03      0.03      0.03      0.03  

[1313 rows x 11 columns]

In [27]:
quiz_data_copy.set_index('Area').loc['Algeria'].groupby('Element')[['Y2018']].sum()

Y2018
Element                                         
Domestic supply quantity                90023.42
Export Quantity                           811.54
Fat supply quantity (g/capita/day)        309.09
Feed                                    10668.00
Food                                    63736.33
Food supply (kcal/capita/day)           13288.00
Food supply quantity (kg/capita/yr)      1509.28
Import Quantity                         36238.29
Losses                                   7708.00
Other uses (non-food)                    3970.08
Processing                               2479.00
Production                              60766.67
Protein supply quantity (g/capita/day)    367.28
Residuals                                 630.00
Seed                                      825.00
Stock Variation                          6172.00
Total Population - Both sexes           42228.00

In [28]:
quiz_data_copy.set_index('Area').loc['Algeria'].groupby('Element')[['Y2018']].sum().loc['Import Quantity']

Y2018    36238.29
Name: Import Quantity, dtype: float64

## Question 20

What is the total number of unique countries in the dataset?

In [29]:
quiz_data['Area'].nunique()

49

In [30]:
quiz_data['Area'].unique()

array(['Algeria', 'Angola', 'Benin', 'Botswana', 'Burkina Faso',
       'Cabo Verde', 'Cameroon', 'Central African Republic', 'Chad',
       'Comoros', 'Congo', "Côte d'Ivoire", 'Djibouti', 'Egypt',
       'Eswatini', 'Ethiopia', 'Ethiopia PDR', 'Gabon', 'Gambia', 'Ghana',
       'Guinea', 'Guinea-Bissau', 'Kenya', 'Lesotho', 'Liberia',
       'Madagascar', 'Malawi', 'Mali', 'Mauritania', 'Mauritius',
       'Morocco', 'Mozambique', 'Namibia', 'Niger', 'Nigeria', 'Rwanda',
       'Sao Tome and Principe', 'Senegal', 'Seychelles', 'Sierra Leone',
       'South Africa', 'Sudan', 'Sudan (former)', 'Togo', 'Tunisia',
       'Uganda', 'United Republic of Tanzania', 'Zambia', 'Zimbabwe'],
      dtype=object)

In [31]:
num_countries = len(quiz_data['Area'].unique())

In [32]:
pd.DataFrame(quiz_data['Area'].unique(), columns=['Country'], index=list(range(1,num_countries+1)))

Country
1                       Algeria
2                        Angola
3                         Benin
4                      Botswana
5                  Burkina Faso
6                    Cabo Verde
7                      Cameroon
8      Central African Republic
9                          Chad
10                      Comoros
11                        Congo
12                Côte d'Ivoire
13                     Djibouti
14                        Egypt
15                     Eswatini
16                     Ethiopia
17                 Ethiopia PDR
18                        Gabon
19                       Gambia
20                        Ghana
21                       Guinea
22                Guinea-Bissau
23                        Kenya
24                      Lesotho
25                      Liberia
26                   Madagascar
27                       Malawi
28                         Mali
29                   Mauritania
30                    Mauritius
31                      Morocco
32                   Mozambique
33                      Namibia
34                        Niger
35                      Nigeria
36                       Rwanda
37        Sao Tome and Principe
38                      Senegal
39                   Seychelles
40                 Sierra Leone
41                 South Africa
42                        Sudan
43               Sudan (former)
44                         Togo
45                      Tunisia
46                       Uganda
47  United Republic of Tanzania
48                       Zambia
49                     Zimbabwe